extensions/GitExtension/FileExplorerGitIntegration/Models/GitLocalRepository.cs

```diff
@@ -0,0 +1,230 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Runtime.InteropServices;
+using LibGit2Sharp;
+using Microsoft.Windows.DevHome.SDK;
+using Serilog;
+using Windows.Foundation.Collections;
+
+namespace FileExplorerGitIntegration.Models;
+
+[ComVisible(false)]
+[ClassInterface(ClassInterfaceType.None)]
+public sealed class GitLocalRepository : ILocalRepository
+{
+    private readonly RepositoryCache? _repositoryCache;
+
+    private readonly Serilog.ILogger log = Log.ForContext("SourceContext", nameof(GitLocalRepository));
+
+    public string RootFolder
+    {
+        get; init;
+    }
+
+    public GitLocalRepository(string rootFolder)
+    {
+        RootFolder = rootFolder;
+    }
+
+    internal GitLocalRepository(string rootFolder, RepositoryCache? cache)
+    {
+        if (!Repository.IsValid(rootFolder))
+        {
+            throw new ArgumentException("Invalid repository path");
+        }
+
+        RootFolder = rootFolder;
+        _repositoryCache = cache;
+    }
+
+    private Repository? OpenRepository()
+    {
+        if (_repositoryCache == null)
+        {
+            log.Debug("Cache is null. Return new repository object");
+            return new Repository(RootFolder);
+        }
+
+        log.Debug("Obtained repository from cache");
+        return _repositoryCache.GetRepository(RootFolder);
+    }
+
+    IPropertySet ILocalRepository.GetProperties(string[] properties, string relativePath)
+    {
+        relativePath = relativePath.Replace('\\', '/');
+        var result = new ValueSet();
+        Commit? latestCommit = null;
+
+        var repository = OpenRepository();
+
+        if (repository == null)
+        {
+            log.Debug("GetProperties: Repository object is null");
+            return result;
+        }
+
+        foreach (var propName in properties)

```

Comment:
I believe the null checks are used because of iterating through the properties and this needs the latest commit to be not null.

I don't like the duplication of the check.

If `LatestCommit` returns null for a property, would a second call to `LatestCommit` return a non-null value?  If this is the case the call to `LatestCommit` can be moved outside the foreach loop.

The null check for `LatestCommit` still needs to be in each case "If not null, add the property"